rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java

```diff
@@ -75,8 +98,190 @@ public NodesLockRestorationManager apply(RMCore rmCore) {
+                nodeStates.put(rmnode.getState(), newNbNodesInState);
+            } else {
+                // the node is not recoverable and does not appear in any data
+                // structures: we can remove it safely from database
+                rmCore.getDbManager().removeNode(rmNodeData);
+                triggerDownNodeHandling(nodeSource, rmNodeData, nodeUrl);
+            }
+            // we must add the recreated node to the eligible data
+            // structure if we want it to be usable by a task
+            if (isEligible(rmnode)) {
+                rmCore.addEligibleNode(rmnode);
+                totalEligibleRecoveredNodes++;
+            }
+        }
+
+        int totalRecoveredNodes = 0;
+
+        logNodeRecoverySummary(nodeSourceName, nodeStates, totalRecoveredNodes, totalEligibleRecoveredNodes);
+    }
+
+    private RMNode recoverNodeInternally(NodeSource nodeSource, RMNodeData rmNodeData, String nodeUrl, Node node) {
+        RMNode rmNode = null;
+        // the node has been successfully looked up, we compare its
+        // information to the node data retrieved in database.
+        if (rmNodeData.equalsToNode(node)) {
+            logger.info("Node to recover could successfully be looked up at URL: " + nodeUrl);
+            rmNode = nodeSource.internalAddNodeAfterRecovery(node, rmNodeData);
+            rmCore.registerAvailableNode(rmNode);
+        } else {
+            logger.error("The node that has been looked up does not have the same information as the node to recover: " +
+                         node.getNodeInformation().getName() + " is not equal to " + rmNodeData.getName() + " or " +
+                         node.getNodeInformation().getURL() + " is not equal to " + rmNodeData.getNodeUrl());
+        }
+        return rmNode;
+    }
+
+    private void triggerDownNodeHandling(NodeSource nodeSource, RMNodeData rmNodeData, String nodeUrl) {
+        // if the node to recover was in deploying state then we have
+        // nothing to do as it is going to be redeployed
+        if (!rmNodeData.getState().equals(NodeState.DEPLOYING)) {
+            // inform the node source that this recreated node is down
+            nodeSource.detectedPingedDownNode(rmNodeData.getName(), nodeUrl);
+        }
+    }
+
+    private Node tryToLookupNode(NodeSource nodeSource, int lookUpTimeout, String nodeUrl) {
+        Node node = null;
+        try {
+            logger.info("Trying to lookup node to recover: " + nodeUrl);
+            node = nodeSource.lookupNode(nodeUrl, lookUpTimeout);
+        } catch (Exception e) {
+            // do not log exception message here: not being able to look up a
+            // node to recover is not an exceptional behavior
+            logger.warn("Node to recover could not be looked up");
+        }
+        return node;
+    }
+
+    private void makeSureNodeSourceHasNoNode(NodeSource nodeSource, String nodeSourceName) {
```

Comment:
can you call this method something which indicates that it logs. 

Like `dispatchLogWarnIfNodeSourceHasNodes`

Reply:
agree